### PR TITLE
Enable creation of software RAID on nodes

### PR DESCRIPTION
**Pull request by jsuchome**
Create extra volumes that should be used for RAID. 
Add first N disks to the RAID array for each node.

Tested with manual mkcloud runs: 
 - worked for raid1 and one node or two nodes
 - also for raid5 and one node

Requires https://github.com/crowbar/crowbar-core/pull/15

### CODE REVIEW

File: scripts/mkcloud [hunk 1]
```diff
@@ -1023,6 +1037,11 @@ Optional
         set new cloudsource for upgrade process
     TESTHEAD='' | 1  (default='')
         use Media from Devel:Cloud:Staging and add test update repositories (except for GM* targets)
+    raidvolumenumber (default=1)
+        The number of disks to join into the software RAID for each node.
+        Default value of 1 means just one system disk and no RAID.
+    raidtype (default='raid11')
```
Comment (dirkmueller) [inline scripts/mkcloud]: raid1

File: scripts/mkcloud [hunk 2]
```diff
@@ -210,6 +212,8 @@ function sshrun()
         export nosetestparameters=${nosetestparameters} ;
         export tempestoptions='${tempestoptions}' ;
         export cephvolumenumber=$cephvolumenumber ;
+        export raidvolumenumber=$raidvolumenumber ;
+        export raidtype=$raidtype ;
```
Comment (bmwiedemann) [inline scripts/mkcloud]: maybe rename raidtype variable to want_raidtype - then it gets auto-forwarded into qa_crowbarsetup. and document it in onadmin_help function and move the default value into qa_crowbarsetup as well, because it has nothing to do with mkcloud script. This allows us to use it for physical clouds.

**Comment by dirkmueller** [comment]
tests seem to fail..

**Comment by jsuchome** [comment]
That is strange, why would ``<target dev='cloud-1'/>`` be in network section?

**Comment by jsuchome** [comment]
@bmwiedemann like this? I wasn't able to test it yet, because of current deployment speed...

File: scripts/mkcloud [hunk 3]
```diff
@@ -1033,6 +1045,11 @@ Optional
         set new cloudsource for upgrade process
     TESTHEAD='' | 1  (default='')
         use Media from Devel:Cloud:Staging and add test update repositories (except for GM* targets)
+    raidvolumenumber (default=1)
+        The number of disks to join into the software RAID for each node.
+        Default value of 1 means just one system disk and no RAID.
+    want_raidtype (default='raid1')
+        The type of RAID to create.
```
Comment (bmwiedemann) [inline scripts/mkcloud]: want_raidtype doc could also be moved into qa_crowbarsetup's onadmin_help function. and squash it into the 3rd commit

File: scripts/lib/libvirt/fixtures/cloud-node1.xml [hunk 4]
```diff
@@ -29,6 +29,7 @@
       <boot order='3'/>
     </disk>
 
+
```
Comment (toabctl) [inline scripts/lib/libvirt/fixtures/cloud-node1.xml]: unrelated

File: scripts/lib/libvirt/compute-config [hunk 5]
```diff
@@ -14,6 +14,7 @@ def main():
                         (when this script is called within a loop, this \
                         is usually an incremented number)")
     parser.add_argument("macaddress", type=str, help="MAC Address")
+    parser.add_argument("raidvolumenumber", type=str, help="Number of disks to create RAID")
```
Comment (toabctl) [inline scripts/lib/libvirt/compute-config]: you can use ```type=int```

File: scripts/lib/libvirt/libvirt_setup.py [hunk 6]
```diff
@@ -106,6 +106,25 @@ def compute_config(args, cpu_flags=cpuflags()):
     else:
         nodememory = args.controllernodememory
 
+    raidvolume = ""
+    if args.raidvolumenumber and args.raidvolumenumber > 1:
+        for i in range(1, int(args.raidvolumenumber)):
```
Comment (toabctl) [inline scripts/lib/libvirt/libvirt_setup.py]: when having ```type=int, default=0``` you can just do ```for i in range(1, args.raidvolumenumber):``` and remove the outer ```if args.raidvolumenumber and args.raidvolumenumber > 1:```

File: scripts/lib/libvirt/libvirt_setup.py [hunk 7]
```diff
@@ -106,6 +106,25 @@ def compute_config(args, cpu_flags=cpuflags()):
...
+            raid_template = string.Template(readfile(
+                "{0}/extra-volume.xml".format(TEMPLATE_DIR)))
+            raid_values = dict(
```
Comment (aplanas) [inline scripts/lib/libvirt/libvirt_setup.py]: this seems strange, can you use the `{}` syntax to build the dict?

File: scripts/lib/libvirt/libvirt_setup.py [hunk 8]
```diff
@@ -106,6 +106,25 @@ def compute_config(args, cpu_flags=cpuflags()):
...
+                volume_serial="{0}-node{1}-raid{2}".format(
+                    args.cloud,
+                    args.nodecounter,
+                    i),
+                source_dev="{0}/{1}.node{2}-raid{3}".format(
+                    args.vdiskdir,
+                    args.cloud,
+                    args.nodecounter,
+                    i),
+                target_dev=targetdevprefix + ''.join(alldevices.next()),
```
Comment (aplanas) [inline scripts/lib/libvirt/libvirt_setup.py]: this `alldevices.next()` inside a `for` looks fishy.  Once the iterator is consumed, this will be empty.

File: scripts/lib/libvirt/fixtures/cloud-node1.xml (hunk 4 above)
Comment (jsuchome) [inline scripts/lib/libvirt/fixtures/cloud-node1.xml]: The empty line is there because raid disks entry is empty for the default test.

File: scripts/lib/libvirt/libvirt_setup.py (hunk 8 above)
Comment (jsuchome) [inline scripts/lib/libvirt/libvirt_setup.py]: It was just copy and past of the ceph disks handling in the same function. Do you think it would need a change as well?
(same for previous comment)

File: scripts/lib/libvirt/fixtures/cloud-node1.xml (hunk 4 above)
Comment (aplanas) [inline scripts/lib/libvirt/fixtures/cloud-node1.xml]: Two end of lines makes a difference in an XML?

File: scripts/lib/libvirt/libvirt_setup.py (hunk 8 above)
Comment (aplanas) [inline scripts/lib/libvirt/libvirt_setup.py]: This is when a test comes handy.  Not sure what do you expect, but the second time that you create this dictionary, the iterator will be empty.  So I guess that there is a bug there and in the source.

Can create some test?

File: scripts/lib/libvirt/fixtures/cloud-node1.xml (hunk 4 above)
Comment (toabctl) [inline scripts/lib/libvirt/fixtures/cloud-node1.xml]: @jsuchome ah. it's for the testcase, right?!

File: scripts/lib/libvirt/compute-config (hunk 5 above)
Comment (bmwiedemann) [inline scripts/lib/libvirt/compute-config]: or default=1 because I think, 0 is not possible.

File: scripts/lib/libvirt/libvirt_setup.py (hunk 7 above)
Comment (bmwiedemann) [inline scripts/lib/libvirt/libvirt_setup.py]: does {} work with python-2.6 ?

Comment (toabctl) [inline scripts/lib/libvirt/libvirt_setup.py]: > does {} work with python-2.6 ?

yes

File: scripts/lib/libvirt/fixtures/cloud-node1.xml (hunk 4 above)
Comment (jsuchome) [inline scripts/lib/libvirt/fixtures/cloud-node1.xml]: yes

File: scripts/lib/libvirt/libvirt_setup.py (hunk 8 above)
Comment (jsuchome) [inline scripts/lib/libvirt/libvirt_setup.py]: It seems that is just the way to pick next letter in alphabet every time device is needed. And based on creation order in mkcloud script, this order is also correct (first raid, than ceph etc.)

**Comment by jsuchome** [comment]
@aplanas , @toabctl good enough?

**Comment by bmwiedemann** [comment]
commits are not the cleanest possible - e.g. changing details about raidvolumenumber in between, but overall the result looks reasonable.

**Comment by aplanas** [comment]
+1

**Comment by toabctl** [comment]
+1
